File: update_wordpress.py
#!/usr/bin/env python3
"""
WordPress Auto-Updater for Julemenyen
Updates WordPress pages with the latest menu from Google Sheets CSV
"""

import os
import sys
import requests
import csv
from io import StringIO
from typing import List, Dict, Optional
import json

# WordPress configuration
WP_URL = "https://julemarkedet-trondheim.no"
WP_USERNAME = "admin"
WP_APP_PASSWORD = os.environ.get("WP_APP_PASSWORD", "")
PAGE_ID_NO = 8498  # Norwegian page
PAGE_ID_EN = 8500  # English page

# CSV URL (published Google Sheets)
CSV_URL = "https://docs.google.com/spreadsheets/d/e/2PACX-1vSo0BYZqtNREhfTCo-ZPWCis5_84iokH3Bqi68itYeRYzjXNsyGK5BWw0omyzHQrwks4vB7yph5-bTk/pub?gid=0&single=true&output=csv"

# GitHub raw URLs for HTML templates
HTML_NO_URL = "https://raw.githubusercontent.com/margaretosoftware/julemenyen/main/html-no.html"
HTML_EN_URL = "https://raw.githubusercontent.com/margaretosoftware/julemenyen/main/html-en.html"


def fetch_csv() -> List[Dict[str, str]]:
    """Fetch and parse CSV from Google Sheets"""
    print(f"📊 Fetching CSV from Google Sheets...")
    response = requests.get(CSV_URL, timeout=10)
    response.raise_for_status()

    csv_text = response.text
    reader = csv.DictReader(StringIO(csv_text))
    rows = [row for row in reader]

    print(f"✅ Fetched {len(rows)} rows from CSV")
    return rows


def get_wordpress_page_content(page_id: int) -> Optional[str]:
    """Get current content of a WordPress page"""
    print(f"📄 Fetching WordPress page {page_id}...")

    url = f"{WP_URL}/wp-json/wp/v2/pages/{page_id}"
    auth = (WP_USERNAME, WP_APP_PASSWORD)

    response = requests.get(url, auth=auth, timeout=10)
    response.raise_for_status()

    data = response.json()
    content = data.get("content", {}).get("rendered", "")

    print(f"✅ Fetched page {page_id} ({len(content)} chars)")
    return content


def update_wordpress_page(page_id: int, new_content: str) -> bool:
    """Update WordPress page content"""
    print(f"📝 Updating WordPress page {page_id}...")

    url = f"{WP_URL}/wp-json/wp/v2/pages/{page_id}"
    auth = (WP_USERNAME, WP_APP_PASSWORD)

    payload = {
        "content": new_content
    }

    response = requests.post(url, auth=auth, json=payload, timeout=10)
    response.raise_for_status()

    print(f"✅ Updated page {page_id}")
    return True


def fetch_html_template(lang: str) -> str:
    """Fetch HTML template from GitHub"""
    url = HTML_NO_URL if lang == "no" else HTML_EN_URL
    print(f"🌐 Fetching {lang.upper()} HTML template from GitHub...")

    response = requests.get(url, timeout=10)
    response.raise_for_status()

    html = response.text
    print(f"✅ Fetched {lang.upper()} template ({len(html)} chars)")
    return html


def normalize_content(content: str) -> str:
    """Normalize HTML content for comparison (remove extra whitespace, etc.)"""
    # Remove extra whitespace and newlines
    normalized = " ".join(content.split())
    return normalized


def needs_update(current_content: str, new_content: str) -> bool:
    """Check if page needs update by comparing normalized content"""
    current_norm = normalize_content(current_content)
    new_norm = normalize_content(new_content)

    # If they're identical, no update needed
    if current_norm == new_norm:
        return False

    # If the lengths differ significantly, update needed
    len_diff = abs(len(current_norm) - len(new_norm))
    if len_diff > 100:  # More than 100 chars difference
        return True

    # Check if content is substantially different
    # (simple check: if more than 5% different)
    max_len = max(len(current_norm), len(new_norm))
    similarity_threshold = 0.95

    # Count matching characters (simplified)
    matches = sum(1 for a, b in zip(current_norm, new_norm) if a == b)
    similarity = matches / max_len if max_len > 0 else 1.0

    return similarity < similarity_threshold


def main():
    """Main execution"""
    print("=" * 60)
    print("🎄 Julemenyen WordPress Auto-Updater")
    print("=" * 60)

    # Check for Application Password
    if not WP_APP_PASSWORD:
        print("❌ ERROR: WP_APP_PASSWORD environment variable not set")
        print("   Set it with: export WP_APP_PASSWORD='your-app-password'")
        sys.exit(1)

    try:
        # Fetch CSV data
        csv_rows = fetch_csv()

        # Process Norwegian page
        print("\n" + "=" * 60)
        print("🇳🇴 Processing Norwegian page...")
        print("=" * 60)

        html_no = fetch_html_template("no")
        current_no = get_wordpress_page_content(PAGE_ID_NO)

        if needs_update(current_no, html_no):
            print("🔄 Norwegian page needs update")
            update_wordpress_page(PAGE_ID_NO, html_no)
        else:
            print("✅ Norwegian page is already up to date")

        # Process English page
        print("\n" + "=" * 60)
        print("🇬🇧 Processing English page...")
        print("=" * 60)

        html_en = fetch_html_template("en")
        current_en = get_wordpress_page_content(PAGE_ID_EN)

        if needs_update(current_en, html_en):
            print("🔄 English page needs update")
            update_wordpress_page(PAGE_ID_EN, html_en)
        else:
            print("✅ English page is already up to date")

        print("\n" + "=" * 60)
        print("✅ Update process completed successfully")
        print("=" * 60)

    except requests.exceptions.RequestException as e:
        print(f"\n❌ ERROR: Network request failed: {e}")
        sys.exit(1)
    except Exception as e:
        print(f"\n❌ ERROR: Unexpected error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: update_wordpress.js
#!/usr/bin/env node

/**
 * WordPress Auto-Updater for Julemenyen
 * Renders pages with JSDOM and uploads to WordPress
 */

const { JSDOM } = require('jsdom');
const https = require('https');
const fs = require('fs');

// WordPress configuration
const WP_URL = 'https://julemarkedet-trondheim.no';
const WP_USERNAME = 'admin';
const WP_APP_PASSWORD = process.env.WP_APP_PASSWORD || '';
const PAGE_ID_NO = 8498;
const PAGE_ID_EN = 8500;

/**
 * WordPress REST API helper
 */
function wpRequest(endpoint, method = 'GET', data = null) {
  return new Promise((resolve, reject) => {
    const auth = Buffer.from(`${WP_USERNAME}:${WP_APP_PASSWORD}`).toString('base64');

    const options = {
      hostname: new URL(WP_URL).hostname,
      path: `/wp-json/wp/v2/${endpoint}`,
      method,
      headers: {
        'Authorization': `Basic ${auth}`,
        'Content-Type': 'application/json',
      }
    };

    const req = https.request(options, (res) => {
      let body = '';
      res.on('data', (chunk) => { body += chunk; });
      res.on('end', () => {
        if (res.statusCode >= 200 && res.statusCode < 300) {
          resolve(JSON.parse(body));
        } else {
          reject(new Error(`HTTP ${res.statusCode}: ${body}`));
        }
      });
    });

    req.on('error', reject);

    if (data && method !== 'GET') {
      req.write(JSON.stringify(data));
    }

    req.end();
  });
}

/**
 * Render page with JSDOM and execute menu.js
 */
async function renderPage(htmlFile) {
  console.log(`🌐 Rendering ${htmlFile}...`);

  // Read HTML file
  const html = fs.readFileSync(htmlFile, 'utf8');

  // Create JSDOM instance
  const dom = new JSDOM(html, {
    runScripts: 'dangerously',
    resources: 'usable',
    url: 'file://' + __dirname + '/' + htmlFile
  });

  const { window } = dom;
  global.window = window;
  global.document = window.document;

  // Wait for menu.js to execute
  await new Promise((resolve) => {
    window.addEventListener('DOMContentLoaded', () => {
      // Give menu.js time to fetch CSV and render
      setTimeout(resolve, 3000);
    });
  });

  // Extract rendered content
  const content = dom.serialize();

  console.log(`✅ Rendered ${htmlFile} (${content.length} chars)`);
  return content;
}

/**
 * Update WordPress page
 */
async function updatePage(pageId, content) {
  console.log(`📝 Updating WordPress page ${pageId}...`);

  try {
    await wpRequest(`pages/${pageId}`, 'POST', { content });
    console.log(`✅ Updated page ${pageId}`);
  } catch (error) {
    console.error(`❌ Error updating page ${pageId}:`, error.message);
    throw error;
  }
}

/**
 * Main execution
 */
async function main() {
  console.log('='.repeat(60));
  console.log('🎄 Julemenyen WordPress Auto-Updater');
  console.log('='.repeat(60));

  if (!WP_APP_PASSWORD) {
    console.error('❌ ERROR: WP_APP_PASSWORD environment variable not set');
    process.exit(1);
  }

  try {
    // Process Norwegian page
    console.log('\n' + '='.repeat(60));
    console.log('🇳🇴 Processing Norwegian page...');
    console.log('='.repeat(60));

    const renderedNO = await renderPage('html-no-local.html');
    await updatePage(PAGE_ID_NO, renderedNO);

    // Process English page
    console.log('\n' + '='.repeat(60));
    console.log('🇬🇧 Processing English page...');
    console.log('='.repeat(60));

    const renderedEN = await renderPage('html-en-local.html');
    await updatePage(PAGE_ID_EN, renderedEN);

    console.log('\n' + '='.repeat(60));
    console.log('✅ Update process completed successfully');
    console.log('='.repeat(60));

  } catch (error) {
    console.error('\n❌ ERROR:', error.message);
    console.error(error.stack);
    process.exit(1);
  }
}

// Run
main();
